management/server/geolocation/store.go

```diff
@@ -0,0 +1,148 @@
+package geolocation
+
+import (
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+	"gorm.io/gorm/logger"
+)
+
+// SqliteStore represents a location storage backed by a Sqlite DB.
+type SqliteStore struct {
+	db       *gorm.DB
+	filePath string
+	mux      *sync.RWMutex
+}
+
+func NewSqliteStore(dataDir string) (*SqliteStore, error) {
+	file := filepath.Join(dataDir, "geonames.db")
+	_, err := fileExists(file)
+	if err != nil {
+		return nil, err
+	}
+
+	db, err := connectDB(file)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SqliteStore{
+		db:       db,
+		filePath: file,
+		mux:      &sync.RWMutex{},
+	}, nil
+}
+
+// GetAllCountries returns a list of all countries in the store.
+func (s *SqliteStore) GetAllCountries() ([]Country, error) {
+	var countries []Country
+	result := s.db.Table("geonames").
+		Select("country_iso_code", "country_name").
+		Group("country_name").
+		Scan(&countries)
+	if result.Error != nil {
+		return nil, result.Error
+	}
+	return countries, nil
+}
+
+// GetCitiesByCountry retrieves a list of cities from the store based on the given country ISO code.
+func (s *SqliteStore) GetCitiesByCountry(countryISOCode string) ([]City, error) {
+	var cities []City
+	result := s.db.Table("geonames").
+		Select("geoname_id", "city_name").
+		Where("country_iso_code = ?", countryISOCode).
+		Group("city_name").
+		Scan(&cities)
+	if result.Error != nil {
+		return nil, result.Error
+	}
+
+	return cities, nil
+}
+
+func (s *SqliteStore) reload() error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	log.Infof("Reloading '%s'", s.filePath)
+
+	db, err := connectDB(s.filePath)
+	if err != nil {
+		return err
+	}
+	s.db = db
```

Comment:
Should we worry about freeing the memory of previously opened db?

Reply:
Yes, I missed it, but it's essential because we loaded it into memory.